internal/app/machined/pkg/controllers/block/lvm.go

```diff
@@ -0,0 +1,172 @@
+type LVMActivationController struct {
+	seenVolumes  map[string]struct{}
+	activatedVGs map[string]struct{}
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *LVMActivationController) Name() string {
+	return "block.LVMActivationController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *LVMActivationController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: v1alpha1.NamespaceName,
+			Type:      runtimeres.MountStatusType,
+			ID:        optional.Some(constants.EphemeralPartitionLabel),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: block.NamespaceName,
+			Type:      block.DiscoveredVolumeType,
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *LVMActivationController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *LVMActivationController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	if ctrl.seenVolumes == nil {
+		ctrl.seenVolumes = make(map[string]struct{})
+	}
+
+	if ctrl.activatedVGs == nil {
+		ctrl.activatedVGs = make(map[string]struct{})
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		if _, err := safe.ReaderGetByID[*runtimeres.MountStatus](ctx, r, constants.EphemeralPartitionLabel); err != nil {
+			if state.IsNotFoundError(err) {
+				// wait for the mount status to be available
+				continue
+			}
+
+			return fmt.Errorf("failed to get mount status: %w", err)
+		}
+
+		discoveredVolumes, err := safe.ReaderListAll[*block.DiscoveredVolume](ctx, r)
+		if err != nil {
+			return fmt.Errorf("failed to list discovered volumes: %w", err)
+		}
+
+		var multiErr error
+
+		for iterator := discoveredVolumes.Iterator(); iterator.Next(); {
+			if _, ok := ctrl.seenVolumes[iterator.Value().Metadata().ID()]; ok {
+				continue
+			}
+
+			if iterator.Value().TypedSpec().Name != "lvm2-pv" {
+				ctrl.seenVolumes[iterator.Value().Metadata().ID()] = struct{}{}
+
+				continue
+			}
+
+			logger.Info("checking device for LVM volume activation", zap.String("device", iterator.Value().TypedSpec().DevPath))
+
+			vgName, err := ctrl.checkVGNeedsActivation(ctx, iterator.Value().TypedSpec().DevPath)
+			if err != nil {
+				multiErr = multierror.Append(multiErr, err)
+
+				continue
+			}
+
+			if vgName == "" {
+				ctrl.seenVolumes[iterator.Value().Metadata().ID()] = struct{}{}
+
+				continue
+			}
+
+			logger.Info("activating LVM volume", zap.String("name", vgName))
+
+			// activate the volume group
+			if _, err = cmd.RunContext(ctx,
+				"/sbin/lvm",
+				"vgchange",
+				"-aay",
+				"--autoactivation",
+				"event",
+				vgName,
+			); err != nil {
+				return fmt.Errorf("failed to activate LVM volume %s: %w", vgName, err)
+			}
+
+			ctrl.activatedVGs[vgName] = struct{}{}
```

Comment:
it looks like we never use this value, do we need to skip activation if already activated?